File: catkin_larc/src/main_engine/src/main_engine.cpp
//ros basic imports
#include <ros/ros.h>
#include <iostream>
#include <std_msgs/Bool.h>
#include <std_msgs/Int32.h>
#include <std_msgs/Float64.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/Twist.h>

#include "vision/objectDetectionArray.h"
#include "vision/objectDetection.h"

using namespace std;

class MainEngine{
    private:
        ros::NodeHandle nh;
        ros::Publisher pub_cmd_vel;
        ros::Publisher pub_intake;
        ros::Publisher pub_elevator;
        ros::Publisher pub_warehouse;
        ros::Publisher pub_reset_odom;
        ros::Publisher pub_global_setpoint;

        ros::Subscriber sub_odom;
        ros::Subscriber sub_color_detect;
        bool target_reached = true;
        pair<float, float> target_position;
        pair<float, float> current_position;
        float tolerance = 0.03;
        int task = 1;
        float imu_setpoint = 0.0;
        int state = 0;

        ros::Time last_spin_time;
        ros::Time state_time;
        bool color_sequence_detected = false;
        bool color_detected[4] = {false, false, false, false};
        int seq_id = 0;
        float detection_max_size = 0;
        float detection_max_y = 0;
        float detection_y_pos[5] = {0, 0, 0, 0, 0};
        string static_color_seq = "gbyrybg";

        geometry_msgs::Twist twist_msg;
        std_msgs::Int32 int_msg;
        std_msgs::Bool bool_msg;

    public:
        MainEngine(){ //constructor
            //init node
            nh = ros::NodeHandle();
            //init publisher
            pub_cmd_vel = nh.advertise<geometry_msgs::Twist>("cmd_vel", 10);
            pub_intake = nh.advertise<std_msgs::Int32>("intake", 10);
            pub_elevator = nh.advertise<std_msgs::Int32>("elevator", 10);
            pub_warehouse = nh.advertise<std_msgs::Int32>("warehouse", 10);
            pub_reset_odom = nh.advertise<std_msgs::Bool>("reset_odom", 5);
            pub_global_setpoint = nh.advertise<std_msgs::Bool>("global_setpoint", 10);

            //init subscriber
            sub_odom = nh.subscribe("odom", 10, &MainEngine::odomCallback, this);
            sub_color_detect = nh.subscribe("color_detect", 20, &MainEngine::colorDetectCallback, this);

            target_reached = true;
            state = 0;

            last_spin_time = ros::Time::now();
        }

        void run(){
            switch( state ){
                case 0:
                    timedSpin(-1);
                    if( color_sequence_detected ){
                        ROS_INFO("0");
                        bool_msg.data = true;
                        pub_global_setpoint.publish(bool_msg);
                        state_time = ros::Time::now();
                        state = 1;
                    } 
                    break;
                case 1:
                    if( ros::Time::now() - state_time >= ros::Duration(3) ){ 
                        ROS_INFO("1");          
                        bool_msg.data = true;
                        pub_reset_odom.publish(bool_msg);

                        target_position = make_pair(0, -0.5);
                        target_reached = false;
                        state = 2;
                    }
                    break;
                case 2:
                    followTarget();
                    if( target_reached ){
                        ROS_INFO("2");
                        twist_msg.linear.x = 0;
                        twist_msg.linear.y = 0;
                        twist_msg.linear.z = 0;
                        twist_msg.angular.x = 0;
                        twist_msg.angular.y = 0;
                        twist_msg.angular.z = 0;
                        pub_cmd_vel.publish(twist_msg);
                        state_time = ros::Time::now();

                        state = 3;
                    }
                    break;
                case 3:
                    if( ros::Time::now() - state_time >= ros::Duration(3) ){                        
                        bool_msg.data = true;
                        pub_reset_odom.publish(bool_msg);
                        int_msg.data = 1;
                        pub_intake.publish(int_msg);

                        target_position = make_pair(-0.2, 0);
                        target_reached = false;
                        state = 4;
                    }
                    break;
            }
        }

        void timedSpin(int dir){
            if( ros::Time::now() - last_spin_time >= ros::Duration(4) ){
                geometry_msgs::Twist msg;
                msg.linear.x = 0;
                msg.linear.y = 0;
                msg.linear.z = 0;
                msg.angular.x = 0;
                msg.angular.y = 0;
                msg.angular.z = 0.5 * dir;
                pub_cmd_vel.publish(msg);
                last_spin_time = ros::Time::now();
            }
        }


        void followTarget(){
            if( !target_reached ){
                if( abs(current_position.first - target_position.first) <= tolerance && abs(current_position.second - target_position.second) <= tolerance ){
                    target_reached = true;
                    //rosinfo target reached
                    ROS_INFO("Target reached");
                    //stop
                    geometry_msgs::Twist msg;
                    msg.linear.x = 0;
                    msg.linear.y = 0;
                    msg.linear.z = 0;
                    msg.angular.x = 0;
                    msg.angular.y = 0;
                    msg.angular.z = 0;
                    pub_cmd_vel.publish(msg);
                }
                else {
                    //move to target
                    //rosinfo target
                    ROS_INFO("Moving to target from current (X,Y): ");
                    ROS_INFO("X: %f", current_position.first);
                    ROS_INFO("Y: %f", current_position.second);

                    geometry_msgs::Twist msg;
                    msg.linear.x = (target_position.first - current_position.first) * 0.7;
                    msg.linear.y = (target_position.second - current_position.second) * 0.7;
                    msg.linear.z = 0;
                    msg.angular.x = 0;
                    msg.angular.y = 0;
                    msg.angular.z = 0;

                    pub_cmd_vel.publish(msg);
                }
            } 
        }

        pair<float, float> readTarget(){
            //rosinfo target
            ROS_INFO("Enter target (X,Y): ");
            float x, y;
            cin>>x;
            cin>>y;
            return make_pair(x, y);
        }

        void odomCallback(const nav_msgs::Odometry::ConstPtr& msg){
            //get current position
            current_position = make_pair(msg->pose.pose.position.x, msg->pose.pose.position.y);
        }

        void colorDetectCallback(const vision::objectDetectionArray::ConstPtr& msg){
            int sz = msg->detections.size();
            if( sz == 0 )
                return;
            float y_min_first = msg->detections[0].ymin;
            float x_last_max = msg->detections[0].xmax;
            string color_seq = "";
            for( int i=1; i<sz; i++ ){
                if( abs( msg->detections[i].ymin - y_min_first ) >= 50 || abs( msg->detections[i].xmin - x_last_max ) >= 50 )
                    break;
                if( msg->detections[i].labelText == "rojo" )
                    color_seq += "r";
                else if( msg->detections[i].labelText == "verde" )
                    color_seq += "g";
                else if( msg->detections[i].labelText == "azul" )
                    color_seq += "b";
                else if( msg->detections[i].labelText == "amarillo" )
                    color_seq += "y";

                x_last_max = msg->detections[i].xmax;
            }

            //check if color sequence is subsequence of static color sequence
            if( static_color_seq.find(color_seq) != string::npos && color_seq.size() >= 2 ){
                color_sequence_detected = true;
            }

            ROS_INFO("colors: %s", color_seq.c_str());
        }

        void manualCommand(){
            char key;
            int command;
            std_msgs::Int32 msg;
            cin>>key;
            cin>>command;

            switch(key){
                case 'Q':
                    msg.data = command;
                    pub_intake.publish(msg);
                    break;
                case 'q':
                    msg.data = command;
                    pub_intake.publish(msg);
                    break;
                case 'E':
                    msg.data = command;
                    pub_elevator.publish(msg);
                    break;
                case 'e':
                    msg.data = command;
                    pub_elevator.publish(msg);
                    break;
                case 'W':
                    msg.data = command;
                    pub_warehouse.publish(msg);
                    break;
                case 'w':
                    msg.data = command;
                    pub_warehouse.publish(msg);
                    break;
                case 'r':
                    std_msgs::Bool msgB;
                    msgB.data = true;
                    pub_reset_odom.publish(msgB);
                    break;
            }
        }
};

int main(int argc, char **argv){
    //init ros
    ros::init(argc, argv, "MainEngine");
    MainEngine mainEngine;

    while (ros::ok()){
        mainEngine.run();
        ros::Rate(10).sleep();
        ros::spinOnce(); 
    }
    return 0;
}

File: catkin_larc/src/main_engine/scripts/mainEngine.py
#!/usr/bin/env python3
#encoding=utf-8
#wait for message

import rospy
from std_msgs.msg import Bool, Int32, Float64, String
from geometry_msgs.msg import Pose, Twist
from nav_msgs.msg import Odometry
from vision.msg import objectDetection, objectDetectionArray
from enum import Enum

#import sys

#sys.path.append(str(pathlib.Path(__file__).parent) + '/../include')

class State(Enum):
    START = 1
    FWD1 = 2
    WAIT1 = 3
    ELE1 = 4
    STACK1 = 5
    LINES_FWD = 6
    OUT_EL = 7
    OUT_WAREHOUSE = 8
    OUT_EL_POS = 9
    OUT_FWD = 10
    OUT_INTAKE = 11
    OUT_WAIT = 12
    OUT_RESET = 13
    ROTATE_180 = 14
    FWD_COLOR = 15
    FINISH = 16

class MainEngine:
    def __init__(self):
        # publishers
        self.pubCmdVel = rospy.Publisher('/cmd_vel', Twist, queue_size=10)
        self.pubIntake = rospy.Publisher('/intake', Int32, queue_size=10)
        self.pubElevator = rospy.Publisher('/elevator', Int32, queue_size=10)
        self.pubWarehouse = rospy.Publisher('/warehouse', Int32, queue_size=10)
        self.pubResetOdom = rospy.Publisher('/reset_odom', Bool, queue_size=10)
        self.pubGlobalSetpoint = rospy.Publisher('/global_setpoint', Bool, queue_size=10)

        # subscribers
        self.subOdom = rospy.Subscriber('/odom', Odometry, self.odomCallback)
        self.subColorDetect = rospy.Subscriber('/color_detect', objectDetectionArray, self.colorDetectCallback)

        # variables
        self.current_time = rospy.Time.now()
        self.state_time = rospy.Time.now()
        self.state = State.START
        self.odom = Odometry()
        self.static_color_seq = "GBYRYBG" # static color sequence
        self.xSquare = 0
        self.color_sequence_detected = False

        self.targetReached = False
        self.targetX = 0
        self.targetY = 0
        self.posTolerance = 0.02
        self.iter = 0
        self.rotateCount = 0

    def odomCallback(self, data):
        self.odom = data

    def colorDetectCallback(self, data):
        sz = len(data.detections)
        if sz == 0:
            return
    
        y_min_first = data.detections[0].ymin
        x_last_max = data.detections[0].xmin
        point_x_min_id = 0
        color_seq = ""
    
        for i in range(sz):
            if abs(data.detections[i].ymin - y_min_first) >= 80 or abs(data.detections[i].xmin - x_last_max) >= 70:
                continue
            if data.detections[i].labelText == "rojo":
                color_seq += "R"
            elif data.detections[i].labelText == "verde":
                color_seq += "G"
            elif data.detections[i].labelText == "azul":
                color_seq += "B"
            elif data.detections[i].labelText == "amarillo":
                color_seq += "Y"

            x_last_max = data.detections[i].xmax
            if( abs(data.detections[i].point3D.x) < abs(data.detections[point_x_min_id].point3D.x)):
                point_x_min_id = i

        #check if subsequence
        #print(color_seq)
        if color_seq in self.static_color_seq and len(color_seq) >= 3:
            rospy.loginfo("Color sequence detected: " + color_seq)
            self.color_sequence_detected = True
            #get square from closer point x and adjacents
            x_square_label = data.detections[point_x_min_id].labelText
            x_square_cont = ""
            if point_x_min_id > 0:
                x_square_cont = data.detections[point_x_min_id - 1].labelText + x_square_label
            else:
                x_square_cont = x_square_label + data.detections[point_x_min_id + 1].labelText

            if x_square_label == "verde" and x_square_cont == "verdeazul":
                self.xSquare = 7
            elif x_square_label == "azul" and (x_square_cont == "verdeazul" or x_square_cont == "azulamarillo"):
                self.xSquare = 6
            elif x_square_label == "amarillo" and (x_square_cont == "azulamarillo" or x_square_cont == "rojoamarillo"):
                self.xSquare = 5
            elif x_square_label == "rojo":
                self.xSquare = 4
            elif x_square_label == "amarillo" and (x_square_cont == "rojoamarillo" or x_square_cont == "amarilloazul"):
                self.xSquare = 3
            elif x_square_label == "azul" and (x_square_cont == "amarilloazul" or x_square_cont == "azulverde"):
                self.xSquare = 2
            elif x_square_label == "verde" and x_square_cont == "azulverde":
                self.xSquare = 1

            print( "x_square: " + str(self.xSquare) )
            

    def driveToTarget(self):
        if abs( self.odom.pose.pose.position.x - self.targetX ) <= self.posTolerance and abs( self.odom.pose.pose.position.y - self.targetY ) <= self.posTolerance:
            self.targetReached = True
            rospy.loginfo("Target reached")
            cmd_vel = Twist()
            cmd_vel.linear.x = 0
            cmd_vel.linear.y = 0
            cmd_vel.linear.z = 0
            cmd_vel.angular.x = 0
            cmd_vel.angular.y = 0
            cmd_vel.angular.z = 0
            self.pubCmdVel.publish(cmd_vel)
        else:
            print( "current X: " +str(self.odom.pose.pose.position.x) + " current Y: " + str(self.odom.pose.pose.position.y) )
            cmd_vel = Twist()
            cmd_vel.linear.x = (self.targetX - self.odom.pose.pose.position.x) * 2.5
            cmd_vel.linear.y = (self.targetY - self.odom.pose.pose.position.y) * 2.5
            cmd_vel.linear.z = 0
            cmd_vel.angular.x = 0
            cmd_vel.angular.y = 0
            cmd_vel.angular.z = 0
            self.pubCmdVel.publish(cmd_vel)


    def run(self):
        self.current_time = rospy.Time.now()

        if self.state == State.START:
            if (self.current_time - self.state_time).to_sec() > 3:
                self.state = State.FWD1
                tx = 0.11 + self.iter*0.11
                self.targetReached = False
                self.setTarget(tx, 0)
                self.pubIntake.publish(1)
                self.state_time = self.current_time
                rospy.loginfo("State changed to FWD1")

        elif self.state == State.FWD1:
            self.driveToTarget()
            if self.targetReached or (self.current_time - self.state_time).to_sec() > 8:
                self.state = State.WAIT1
                self.state_time = self.current_time
                rospy.loginfo("State changed to FWD2")

        elif self.state == State.WAIT1:
            if (self.current_time - self.state_time).to_sec() > 2:
                self.state = State.ELE1
                if self.iter == 0 or self.iter == 2:
                    self.pubElevator.publish(3)
                elif self.iter == 1 or self.iter == 3:
                    self.pubElevator.publish(2)
                self.state_time = self.current_time
                rospy.loginfo("State changed to FWD1")

        elif self.state == State.ELE1:
            if (self.current_time - self.state_time).to_sec() > 3:
                self.state = State.STACK1
                self.pubIntake.publish(2)
                self.state_time = self.current_time
                rospy.loginfo("State changed to SET_SETPOINT")

        elif self.state == State.STACK1:
            if (self.current_time - self.state_time).to_sec() > 3:
                self.pubElevator.publish(1)
                self.state_time = self.current_time
                self.iter += 1
                if self.iter < 4:
                    self.state = State.START
                else:
                    self.state = State.LINES_FWD
                    self.targetReached = False
                    self.setTarget(0.7, 0)
                    self.iter = 0
                rospy.loginfo("State changed to START")
        
        elif self.state == State.LINES_FWD:
            self.driveToTarget()
            if self.targetReached or (self.current_time - self.state_time).to_sec() > 10:
                self.pubElevator.publish(5)
                self.state = State.OUT_EL
                rospy.loginfo("State changed to LINE FOLLOW")

        elif self.state == State.OUT_EL:
            if (self.current_time - self.state_time).to_sec() > 4:
                self.state_time = self.current_time
                if self.iter == 0 or self.iter == 2:
                    self.pubWarehouse.publish(3)
                elif self.iter == 1 or self.iter ==3:
                    self.pubWarehouse.publish(2)
                self.state = State.OUT_WAREHOUSE
                rospy.loginfo("State changed to WAREHOUSE")
        
        elif self.state == State.OUT_WAREHOUSE:
            if (self.current_time - self.state_time).to_sec() > 4:
                self.state_time = self.current_time
                self.pubIntake.publish(3)
                self.state = State.OUT_EL_POS
                rospy.loginfo("State changed to ELEVATOR PLACE")

        elif self.state == State.OUT_EL_POS:
            if (self.current_time - self.state_time).to_sec() > 4:
                self.state_time = self.current_time
                if self.iter == 0:
                    self.pubElevator.publish(8)
                if self.iter == 1:
                    self.pubElevator.publish(9)
                if self.iter == 2:
                    self.pubElevator.publish(10)
                if self.iter == 3:
                    self.pubElevator.publish(1)
                self.state = State.OUT_FWD
                rospy.loginfo("State changed to FWD")
        
        elif self.state == State.OUT_FWD:
            if (self.current_time - self.state_time).to_sec() > 4:
                self.state_time = self.current_time
                if self.iter == 3:
                    self.pubIntake.publish(4)
                    self.state = State.FINISH
                    rospy.loginfo("State changed to FINISH")
                else:
                    self.targetReached = False
                    self.setTarget(0.85, 0)
                    self.state = State.OUT_INTAKE
                    rospy.loginfo("State changed to INTAKE")

        elif self.state == State.OUT_INTAKE:
            self.driveToTarget()
            if self.targetReached or (self.current_time - self.state_time).to_sec() > 8:
                self.state_time = self.current_time
                self.pubIntake.publish(4)
                self.state = State.OUT_WAIT
                rospy.loginfo("State changed to OUT WAIT")
        
        elif self.state == State.OUT_WAIT:
            if (self.current_time - self.state_time).to_sec() > 4:
                self.state_time = self.current_time
                self.targetReached = False
                self.setTarget(0.7, 0)
                self.state = State.OUT_RESET
                rospy.loginfo("State changed to RESET")

        elif self.state == State.OUT_RESET:
            self.driveToTarget()
            if self.targetReached or (self.current_time - self.state_time).to_sec() > 8:
                self.state_time = self.current_time
                self.iter += 1
                self.state = State.OUT_EL
                if self.iter == 1:
                    self.pubElevator.publish(6)
                elif self.iter == 2:
                    self.pubElevator.publish(5)
                elif self.iter >= 3:
                    self.rotateCount = 0
                    self.state = State.FINISH
                rospy.loginfo("State changed to OUT ROTATE")

        elif self.state == State.ROTATE_180:
            if (self.current_time - self.state_time).to_sec() > 5:
                self.rotateCount += 1
                if( self.rotateCount >= 3 ):
                    self.targetReached = False
                    self.setTarget(1.7, 0)
                    self.state = State.FWD_COLOR
                    rospy.loginfo("State changed to COLOR FWD")
                else:
                    self.driveSpin(1)
                self.state_time = self.current_time

        elif self.state == State.FWD_COLOR:
            self.driveToTarget()
            if self.targetReached or (self.current_time - self.state_time).to_sec() > 15:
                self.state_time = self.current_time
                self.pubIntake.publish(6)
                self.state = State.OUT_EL
                rospy.loginfo("State changed to OUT EL COLOR")


        #jasdkjasdjkasd
    
    def setTarget(self, x, y):
        self.targetX = x
        self.targetY = y

    def driveSpin(self, dir):
        #publish to cmd_vel with angular velocity Z = dir * 0.5
        cmd_vel = Twist()
        cmd_vel.angular.z = -dir * 0.5
        self.pubCmdVel.publish(cmd_vel)

    def driveFwdToLine(self):
        #publish to cmd_vel with linear velocity X = 0.5
        cmd_vel = Twist()
        cmd_vel.linear.y = 0.4
        self.pubCmdVel.publish(cmd_vel)
    
    def driveFwdTimed(self):
        cmd_vel = Twist()
        cmd_vel.linear.x = 0.4
        self.pubCmdVel.publish(cmd_vel)

if __name__ == '__main__':
    rospy.init_node('mainEngine', anonymous=True)
    rospy.loginfo("mainEngine node started")
    rate = rospy.Rate(10) # 10hz
    
    try:
        mainEngine = MainEngine()
    except rospy.ROSInterruptException:
        pass

    while not rospy.is_shutdown():
        mainEngine.run()   
        rate.sleep()